### PR TITLE
Feature/co 840 post

### PR DESCRIPTION
**Pull request by jaredkosanovic**

### CODE REVIEW

File: src/main/groovy/edu/oregonstate/mist/inventory/AllowedValues.groovy [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package edu.oregonstate.mist.inventory
+
+class AllowedValues {
+
+    List<String> list
+
+    AllowedValues(List<String> allowedTypes) {
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/AllowedValues.groovy]: We can probably use the default constructor here instead of creating a new constructor.

File: src/main/groovy/edu/oregonstate/mist/inventory/AllowedValues.groovy [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        this.list = allowedTypes
+    }
+
+    /**
+     * Create english sentence from list of strings.
+     * @return
+     */
+    public String pretty() {
+        String prettyString = ""
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/AllowedValues.groovy]: I would make this a bit more generic and simpler. Groovy has some shortcuts, you can use. For example, if your list is "foo":

foo[0..-2].join(", ") + " or " + foo[-1]

Let's also write some unit tests for this method.

File: src/main/groovy/edu/oregonstate/mist/inventory/ErrorMessages.groovy [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package edu.oregonstate.mist.inventory
+
+import edu.oregonstate.mist.api.Error
+
+class ErrorMessages {
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/ErrorMessages.groovy]: This is a good pattern. Given that you have quite a few different types of error messages, how about if we use the resource.properties file for the strings?

File: src/main/groovy/edu/oregonstate/mist/inventory/InventoryApplication.groovy [hunk 4]
```diff
@@ -27,9 +29,13 @@ class InventoryApplication extends Application<InventoryConfiguration> {
         DBIFactory factory = new DBIFactory()
         DBI jdbi = factory.build(environment, configuration.getDataSourceFactory(), "jdbi")
         InventoryDAO inventoryDAO = jdbi.onDemand(InventoryDAO.class)
-        InventoryDAOWrapper inventoryDAOWrapper = new InventoryDAOWrapper(inventoryDAO)
-        environment.jersey().register(new InventoryResource(inventoryDAOWrapper,
-                configuration.api.endpointUri))
+        URI inventorySelfLink = UriBuilder.fromUri(
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/InventoryApplication.groovy]: I wonder if we should have a method that returns the resourceSelfLink, but at the skeleton level for the parent class, Application?

File: src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy [hunk 5]
```diff
@@ -1,19 +1,25 @@
...
 
 class InventoryDAOWrapper {
     private InventoryDAO inventoryDAO
+    private URI selfLinkBase
 
     private final String QUERY_DB_TYPE = "Query"
     private final String PROVIDED_DATA_DB_TYPE = "Provided Data"
 
-    InventoryDAOWrapper(InventoryDAO inventoryDao) {
+    InventoryDAOWrapper(InventoryDAO inventoryDao, URI selfLinkBase) {
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy]: How about if use the default map constructor? We then don't have to maintain this constructor / code.

File: src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy [hunk 6]
```diff
@@ -22,12 +28,12 @@ class InventoryDAOWrapper {
...
      * @return
      */
-    public ResourceObject getInventoryById(String inventoryID, URI selfLinkBase) {
+    public ResourceObject getInventoryById(String inventoryID) {
         Inventory inventory = inventoryDAO.getInventoryByID(inventoryID)
         ResourceObject completeInventory
 
         if (inventory) {
-            completeInventory = inventoryBuilder(inventory, selfLinkBase)
+            completeInventory = inventoryBuilder(inventory)
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy]: We can probably get rid of the temporary assignment & variable:
if (inventory) {
inventoryBuilder(inventory)
} else {
null
}

File: src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy [hunk 7]
```diff
@@ -38,12 +44,12 @@ class InventoryDAOWrapper {
      * @param selfLinkBase
      * @return
      */
-    public List<ResourceObject> getAllInventories(URI selfLinkBase) {
+    public List<ResourceObject> getAllInventories() {
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy]: Using a bit of groovy, we can simplify this method to one line:

inventoryDAO.getInventories().collect { inventoryBuilder(it) }

File: src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy [hunk 8]
```diff
@@ -56,14 +62,21 @@ class InventoryDAOWrapper {
      * @param selfLinkBase
      * @return
      */
-    private ResourceObject inventoryBuilder(Inventory inventory, URI selfLinkBase) {
-        inventory.apiQueryParams = inventoryDAO.getFields(QUERY_DB_TYPE, inventory.id)
+    private ResourceObject inventoryBuilder(Inventory inventory) {
+        inventory.apiQueryParams = inventoryDAO.getFields(
+                QUERY_DB_TYPE,
+                inventory.id,
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy:54]: Is passing inventory.id twice to the same method a typo? If it's not, how about if we add a comment to help with readability later on?

File: src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy (hunk 7 above)
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy]: If we use mockito or another mocking framework to mock: inventoryDAO.getFields(), we should be able to write a unit tests for this method.

File: src/main/groovy/edu/oregonstate/mist/inventory/resources/InventoryResource.groovy [hunk 9]
```diff
@@ -1,65 +1,79 @@
...
 class InventoryResource extends Resource {
 
     Logger logger = LoggerFactory.getLogger(InventoryResource.class)
 
     private InventoryDAOWrapper inventoryDAOWrapper
-    private URI endpointUri
 
-    InventoryResource(InventoryDAOWrapper inventoryDAOWrapper, URI endpointUri) {
+    // used to check if client-generated id is a valid UUID
+    private final String uuidRegEx =
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/resources/InventoryResource.groovy]: Should we move the uuidRegEx to the skeleton?

File: src/main/groovy/edu/oregonstate/mist/inventory/resources/InventoryResource.groovy [hunk 10]
```diff
@@ -69,17 +83,41 @@ class InventoryResource extends Resource {
...
+        if (errors) {
+            Response.ResponseBuilder responseBuilder = Response.status(Response.Status.BAD_REQUEST)
+            return responseBuilder.entity(errors).build()
+        }
+        Inventory newInventory = resultObjectToInventory(newResultObject)
+        newInventory.id = newResultObject.data['id'] ?: randomUUID() as String
+
+        inventoryDAOWrapper.createInventory(newInventory)
+
+        ResultObject newCreatedInventory = new ResultObject(
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/resources/InventoryResource.groovy]: How about checking that inventoryDAOWrapper.getInventoryById(newInventory.id) is not null before building the ResultObject? You can then do some error handling.

**Review by floppyears — CHANGES_REQUESTED** [review]

File: src/main/groovy/edu/oregonstate/mist/inventory/InventoryApplication.groovy (hunk 4 above)
Comment (jaredkosanovic) [reply src/main/groovy/edu/oregonstate/mist/inventory/InventoryApplication.groovy]: I like the idea of having it as a method, but not in the application class. I put it in the configuration class. Let me know what you think :)

File: src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy (hunk 8 above)
Comment (jaredkosanovic) [reply src/main/groovy/edu/oregonstate/mist/inventory/db/InventoryDAOWrapper.groovy:54]: Not a typo. Since fields (whether provided or being used in a query) are stored in the same table, the field record in the DB has a parent ID and the related inventory ID. For api query params, the parentID is also the inventoryID. I updated the method comment in the DAO interface to make this more clear.

File: src/main/groovy/edu/oregonstate/mist/inventory/resources/InventoryResource.groovy (hunk 9 above)
Comment (jaredkosanovic) [reply src/main/groovy/edu/oregonstate/mist/inventory/resources/InventoryResource.groovy]: Yes

File: src/main/groovy/edu/oregonstate/mist/inventory/ErrorMessages.groovy [hunk 11]
```diff
@@ -4,48 +4,49 @@ import edu.oregonstate.mist.api.Error
 
 class ErrorMessages {
     static Error invalidUUID() {
-        Error.badRequest("ID is not a valid UUID. " +
-                "Event ID must follow UUID structure detailed here: " +
-                "https://tools.ietf.org/html/rfc4122.html")
+        Error.badRequest(Error.prop.getProperty('inventory.invalidUUID'))
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/ErrorMessages.groovy]: One small nitpick comment in this class is that you still have 5+ methods that call Error.badRequest(Error.prop.getProperty( String param) ) . You could have one small private method that can provide this functionality, and helps make the code DRYer. Should this class or some of these errors be moved to the skeleton? For example the invalidUUID and idExists would be common for CRUD operations.

File: src/test/groovy/edu/oregonstate/mist/inventory/InventoryDAOWrapperTest.groovy [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        dataSources.add(new DataSource(
+                sourceID: "1",
+                source: "Some huge data warehouse",
+                sourceDescription: "Just a huge data warehouse",
+                sourceType: "DB",
+                internal: true
+        ))
+        stubDAO.demand.getProvidedData() { dataSources }
+
+        InventoryDAO inventoryDAO = stubDAO.proxyInstance()
```
Comment (floppyears) [inline src/test/groovy/edu/oregonstate/mist/inventory/InventoryDAOWrapperTest.groovy]: Good use of mocking. This test method does a lot of work to build the mock stubDAO object. How about creating a private method that takes care of building the stubDAO and returns it so that this test can use it?

File: src/main/resources/edu/oregonstate/mist/api/resource.properties [hunk 13]
```diff
@@ -21,3 +21,15 @@ conflict.details = https://developer.oregonstate.edu/documentation/error-referen
 internalServerError.userMessage = The application ran into an internal error. If the problem persists, please contact application support.
 internalServerError.code = 1500
 internalServerError.details = https://developer.oregonstate.edu/documentation/error-reference#1500
+
+#Inventory API Error Messages
+inventory.invalidUUID = ID is not a valid UUID. Inventory ID must follow UUID structure detailed here: https://tools.ietf.org/html/rfc4122.html
```
Comment (floppyears) [inline src/main/resources/edu/oregonstate/mist/api/resource.properties]: I think the invalidUUID and idExists messages are more generic user messages that are probably not specific to inventory-api.

File: src/test/groovy/edu/oregonstate/mist/inventory/AllowedValuesTest.groovy [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+class AllowedValuesTest {
+
+    private AllowedValues testValues = new AllowedValues(list: ['foo', 'bar', 'eggplant'])
+
+    @Test
+    public void testPrettyString() {
+        String prettyString = testValues.pretty()
+        List<String> parsedPrettyString = prettyString.tokenize(',')
+
+        testValues.list.each {
```
Comment (floppyears) [inline src/test/groovy/edu/oregonstate/mist/inventory/AllowedValuesTest.groovy]: What about testing edge cases? What if there is only 0, 1 or 2 items in AllowedValues?

File: src/main/groovy/edu/oregonstate/mist/inventory/ErrorMessages.groovy (hunk 11 above)
Comment (jaredkosanovic) [reply src/main/groovy/edu/oregonstate/mist/inventory/ErrorMessages.groovy]: I addressed this in a new commit.

File: src/main/resources/edu/oregonstate/mist/api/resource.properties (hunk 13 above)
Comment (jaredkosanovic) [reply src/main/resources/edu/oregonstate/mist/api/resource.properties]: I addressed this in a new commit.

File: src/test/groovy/edu/oregonstate/mist/inventory/AllowedValuesTest.groovy (hunk 14 above)
Comment (jaredkosanovic) [reply src/test/groovy/edu/oregonstate/mist/inventory/AllowedValuesTest.groovy]: Good idea! I added better functionality to the AllowedValues class and added more tests

File: src/test/groovy/edu/oregonstate/mist/inventory/InventoryDAOWrapperTest.groovy (hunk 12 above)
Comment (jaredkosanovic) [reply src/test/groovy/edu/oregonstate/mist/inventory/InventoryDAOWrapperTest.groovy]: Done

File: src/main/groovy/edu/oregonstate/mist/inventory/resources/InventoryResource.groovy [hunk 15]
```diff
@@ -203,12 +227,62 @@ class InventoryResource extends Resource {
...
     }
 
     /**
+     * Returns true if a Field in a List<Field> contains a null fieldID
+     * @param fields
+     * @return
+     */
+    private Boolean nullFieldID(List<Field> fields) {
+        fields.find {
+            if (!it.fieldID) {
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/inventory/resources/InventoryResource.groovy]: These methods are not DRY. You could write a closure that uses a variable if it."$fieldName" to check if something is not set. SO: https://stackoverflow.com/questions/7297696/groovy-have-a-field-name-need-to-set-value-and-dont-want-to-use-switch
